File: AtCoder/Beginner363/E/main.cpp
#include <iostream>
#include <vector>
#include <set>
#include <algorithm>
#include <functional>

using namespace std;

struct Cell {
    int height;
    int i;
    int j;
    Cell(int h, int x, int y) : height(h), i(x), j(y) {}
};

struct CompareCell : public binary_function<Cell, Cell, bool> {
    bool operator()(const Cell& lhs, const Cell& rhs) const {
        return lhs.height > rhs.height;
    }
};

int main() {
    int h, w, y;
    cin >> h >> w >> y;

    vector<vector<int> > grid(h, vector<int>(w));
    for (int i = 0; i < h; ++i) {
        for (int j = 0; j < w; ++j) {
            cin >> grid[i][j];
        }
    }

    vector<Cell> heap;
    set<pair<int, int> > visited;

    for (int i = 0; i < h; ++i) {
        for (int j = 0; j < w; ++j) {
            if (i == 0 || j == 0 || i == h-1 || j == w-1) {
                heap.push_back(Cell(grid[i][j], i, j));
                visited.insert(make_pair(i, j));
            }
        }
    }

    make_heap(heap.begin(), heap.end(), CompareCell());
    int floating = h * w;

    for (int height = 1; height <= y; ++height) {
        while (!heap.empty() && heap.front().height <= height) {
            Cell cell = heap.front();
            pop_heap(heap.begin(), heap.end(), CompareCell());
            heap.pop_back();

            int di[] = {-1, 1, 0, 0};
            int dj[] = {0, 0, -1, 1};

            for (int k = 0; k < 4; ++k) {
                int ni = cell.i + di[k];
                int nj = cell.j + dj[k];

                if (ni >= 0 && ni < h && nj >= 0 && nj < w && 
                    visited.find(make_pair(ni, nj)) == visited.end()) {
                    heap.push_back(Cell(grid[ni][nj], ni, nj));
                    push_heap(heap.begin(), heap.end(), CompareCell());
                    visited.insert(make_pair(ni, nj));
                }
            }

            --floating;
        }
        cout << floating << endl;
    }

    return 0;
}

File: AtCoder/Beginner363/E/main.py
from heapq import heappop, heappush, heapify


h, w, y = map(int, input().split())
grid = [list(map(int, input().split())) for _ in range(h)]
heap = [(grid[i][j], (i, j)) for i in range(h) for j in range(w) if i == 0 or j == 0 or j == w - 1 or i == h - 1]
heapify(heap) 
visited = set([(i, j) for i in range(h) for j in range(w) if i == 0 or j == 0 or j == w - 1 or i == h - 1])
floating = h * w

for height in range(1, y + 1):

    while heap and heap[0][0] <= height:
        _, (i, j) = heappop(heap)
        indexes_around = []
        
        if i > 0 and (i - 1, j) not in visited:
            indexes_around.append((i - 1, j))
        if i < len(grid) - 1 and (i + 1, j) not in visited:
            indexes_around.append((i + 1, j))
        if j > 0 and (i, j - 1) not in visited:
            indexes_around.append((i, j - 1))
        if j < len(grid[0]) - 1 and (i, j + 1) not in visited:
            indexes_around.append((i, j + 1))
        for x, y in indexes_around:
            heappush(heap, (grid[x][y], (x, y)))
            visited.add((x, y))
        floating -= 1
    print(floating)





